apps/ejabberd/src/http_upload/mod_http_upload.erl

```diff
@@ -0,0 +1,128 @@
+    mod_disco:unregister_feature(Host, ?NS_HTTP_UPLOAD).
+
+
+-spec iq_handler(From :: ejabberd:jid(), To :: ejabberd:jid(), IQ :: ejabberd:iq()) ->
+                        ejabberd:iq() | ignore.
+iq_handler(_From, _To, IQ = #iq{type = set, sub_el = SubEl}) ->
+    IQ#iq{type = error, sub_el = [SubEl, ?ERR_NOT_ALLOWED]};
+iq_handler(_From, _To, IQ = #iq{type = get, sub_el = Request}) ->
+    case parse_request(Request) of
+        {Filename, Size, ContentType} ->
+            UTCDateTime = calendar:universal_time(),
+            Token = generate_token(),
+            BackendModule = backend_module(),
+            Opts = module_opts(),
+
+            {PutUrl, GetUrl} = BackendModule:create_slot(UTCDateTime, Token, Filename,
+                                                         ContentType, Size, Opts),
+
+            compose_iq_reply(IQ, PutUrl, GetUrl);
+
+        bad_request ->
+            IQ#iq{type = error, sub_el = [Request, ?ERR_BAD_REQUEST]}
+    end.
+
+%%--------------------------------------------------------------------
+%% Helpers
+%%--------------------------------------------------------------------
+
+-spec compose_iq_reply(IQ :: ejabberd:iq(), PutUrl :: binary(), GetUrl :: binary()) ->
+                              Reply :: ejabberd:iq().
+compose_iq_reply(IQ, PutUrl, GetUrl) ->
+    Slot = #xmlel{
+              name = <<"slot">>,
+              attrs=[{<<"xmlns">>, ?NS_HTTP_UPLOAD}],
+              children =[
+                         #xmlel{name = <<"put">>, children = [exml:escape_cdata(PutUrl)]},
+                         #xmlel{name = <<"get">>, children = [exml:escape_cdata(GetUrl)]}
+                        ]},
+
+    IQ#iq{type = result, sub_el =[Slot]}.
+
+
+-spec token_bytes() -> pos_integer().
+token_bytes() ->
+    gen_mod:get_module_opt(?MYNAME, ?MODULE, token_bytes, 32).
+
+
+-spec backend_module() -> atom().
+backend_module() ->
+    Backend = gen_mod:get_module_opt(?MYNAME, ?MODULE, backend, s3),
+    binary_to_atom(<<"http_upload_", (atom_to_binary(Backend, latin1))/binary>>, latin1).
+
+
+-spec module_opts() -> proplists:proplist().
+module_opts() ->
+    gen_mod:get_module_opts(?MYNAME, ?MODULE).
+
+
+-spec generate_token() -> binary().
+generate_token() ->
+    base16:encode(crypto:strong_rand_bytes(token_bytes())).
+
+
+-spec parse_request(Request :: exml:element()) ->
+                           {Filename :: binary(), Size :: integer(),
+                            ContentType :: binary() | undefined} |
+                           bad_request.
+parse_request(Request) ->
+    Filename = exml_query:path(Request, [{element, <<"filename">>}, cdata]),
+    SizeBin = exml_query:path(Request, [{element, <<"size">>}, cdata]),
+    Size = (catch erlang:binary_to_integer(SizeBin)),
+    ContentType = exml_query:path(Request, [{element, <<"content-type">>}, cdata]),
+
+    case is_binary(Filename) andalso <<>> =/= Filename
+        andalso is_integer(Size) andalso Size >= 0 of
```

Comment:
IMO the server should allow setting max file size in `mod_http_upload` options and enforce it.